src/authentication/authentication-hapi-plugin.ts

```diff
@@ -385,42 +390,45 @@ class AuthenticationHapiPlugin extends HapiPlugin {
     }
   }
 
-  private userHasAccessToTeam(userNameOrRequest: Hapi.Request | string) {
-    return async (teamId: number, shouldThrowOnFalse = false) => {
+  private userHasAccessToTeamDecorator(request: Hapi.Request) {
+    return async(teamId: number) => {
       try {
-        // the username field is set by the 'validateFuncFactory'
-        const userName = typeof userNameOrRequest === 'string' ?
-          userNameOrRequest :
-          userNameOrRequest.auth.credentials.username;
-        return (await this.gitlab.getGroup(teamId, userName)).id === teamId;
+        return this.userHasAccessToTeam(this.getUserName(request), teamId);
       } catch (error) {
-        this.logger.warn(`User not authorized to access team ${teamId}`, error);
-        if (shouldThrowOnFalse) {
-          throw error;
-        }
+        this.logger.warn(`Can't check authorization since user is not authenticated`, error);
         return false;
       }
     };
   }
 
-  private userHasAccessToProject(userNameOrRequest: Hapi.Request | string) {
-    return async (projectId: number, shouldThrowOnFalse = false) => {
+  private userHasAccessToProjectDecorator(request: Hapi.Request) {
+    return async (projectId: number) => {
       try {
-        // the username field is set by the 'validateFuncFactory'
-        const userName = typeof userNameOrRequest === 'string' ?
-          userNameOrRequest :
-          userNameOrRequest.auth.credentials.username;
-        return (await this.gitlab.getProject(projectId, userName)).id === projectId;
+        return this.userHasAccessToProject(this.getUserName(request), projectId);
       } catch (error) {
-        this.logger.warn(`User not authorized to access project ${projectId}`, error);
-        if (shouldThrowOnFalse) {
-          throw error;
-        }
+        this.logger.warn(`Can't check authorization since user is not authenticated`, error);
         return false;
       }
     };
   }
 
+  public async userHasAccessToTeam(userName: string, teamId: number) {
+    try {
+      return (await this.gitlab.getGroup(teamId, userName)) !== undefined;
```

Comment:
Actually, coming to think of it, we should make sure that we don't cache `false` return values resulting from GitLab not being reachable or other "unexpected" problems. 

Maybe the best approach would be not to cache any `false` return values for the access check functions.  Not sure whether that can be done with `memoizee`.

Some sanity check for GitLab response seems useful. It could be possible that a request is not routed properly to a GitLab response, and some other server responds with `200`.